### ISSUE TITLE
`manage_deps()` doesn't respect pinned version

### ISSUE THREAD

**Issue by andrewGhazi** [carpentries/sandpaper#628]
Hi there, I maintain a [new module](https://github.com/carpentries-incubator/bioc-scrnaseq) in the Carpentries incubator. I used `sandpaper::pin_version("duckdb@1.1.2")` in an attempt to pin the dependency on `duckdb` to version 1.1.2. You can see the updated `renv.lock` file [here](https://github.com/carpentries-incubator/bioc-scrnaseq/blob/main/renv/profiles/lesson-requirements/renv.lock#L1816). The build process that gets triggered [shows the dependency on the right version](https://github.com/carpentries-incubator/bioc-scrnaseq/actions/runs/12432764582/job/34712886752#step:6:280) but then ignores that and [tries to install the latest bugged version when setting up the cache to build the site](https://github.com/carpentries-incubator/bioc-scrnaseq/actions/runs/12432764582/job/34712886752#step:6:836), causing the whole build to fail. The subsequent re-attempts also fail from the same issue. 

I believe the offending `sandpaper::manage_deps()` call in the github workflow is [this line](https://github.com/carpentries/actions/blob/main/setup-lesson-deps/action.yaml#L130). I've tried to understand the source code for both the github action and `sandpaper::manage_deps()` but there's just too many layers of abstraction for me to identify the source of the error.

Sidenote: [The failure from duckdb](https://github.com/carpentries-incubator/bioc-scrnaseq/actions/runs/12432764582/job/34712886752#step:6:2591) has [something to do with a cleanup script](https://github.com/duckdb/duckdb-r/issues/612) in the latest version of duckdb which is why I'm trying to pin to 1.1.2.

**Comment by froggleston** [comment]
Hi @andrewGhazi - I'm currently on holiday, and not very knowledgeable in the internals and intricacies of renv, but I'm wondering if it's down to a dependent package pulling in a newer version?

e.g. CuratedAtlasQueryR also depends on duckdb, so I don't know if manage_deps() takes the pin into account if another package also pulls in a newer version...

**Comment by andrewGhazi** [comment]
That seems plausible, but it seems like there must be a way to solve it. It's a pretty frustrating situation because I had thought the entire purpose of all this renv/CI/build infrastructure was to be able to set specific dependency versions.

**Comment by froggleston** [comment]
I agree - I'll be able to take a look at this in the new year most likely, but I'll probably need some community assistance as renv isn't my strong suit.

**Comment by froggleston** [comment]
Hi @andrewGhazi - I've managed to narrow down the offending part of the script: https://github.com/carpentries/actions/blob/c90283f3bc25fa13d80d56bb6cfc7325ada6d995/setup-lesson-deps/action.yaml#L118

This line will always install all requisite dependencies, regardless of version, so whatever you specify in renv.lock will be overwritten with the latest packages available. I'm thinking about ways to solve this.

Thank you for your patience!

**Comment by froggleston** [comment]
Hi again @andrewGhazi - could you try adding an environment variable to your repo? @zkamvar gave me us a useful suggestion!

- add the `R_REMOTES_UPGRADE` variable to your repo environment variables in your settings menu for your lesson
- set the value to this variable to `never`

This should stop `remotes::install_deps()` upgrading your packages that you've pinned in your lockfile. Could you try this out and feed back if this works OK? We could then add it to the sandpaper documentation!

**Comment by andrewGhazi** [comment]
Assuming that I applied that environment variable correctly (the place to apply it is main repo page > Settings tab > Environments > github-pages > Environment variables , right?), it still ignores it.  The same failure mode as before, where it [acknowledges the correct dependency initially](https://github.com/carpentries-incubator/bioc-scrnaseq/actions/runs/12770220416/job/35594553444#step:6:280) but then [tries to pull the latest version from CRAN](https://github.com/carpentries-incubator/bioc-scrnaseq/actions/runs/12770220416/job/35594553444#step:6:836). One other detail that might be relevant is that there is no cached version to upgrade from at this point (I cleared the cache in an attempt to fix this).

**Comment by zkamvar** [comment]
> Assuming that I applied that environment variable correctly (the place to apply it is main repo page > Settings tab > Environments > github-pages > Environment variables , right?), it still ignores it.

That's not quite right. You need to set the environment variable in the workflow directly.

You can add it under line 32: https://github.com/carpentries-incubator/bioc-scrnaseq/actions/runs/12770220416/workflow#L32

```yaml
    env:
      R_REMOTES_UPGRADE: never
```

**Comment by froggleston** [comment]
I found this issue whilst searching for insights: https://github.com/r-lib/remotes/issues/815

And this has me worried - does upgrade = "never" have some undocumented exclusions that it only works with CRAN packages? Or am I misunderstanding?

**Comment by froggleston** [comment]
This is now resolved in the later workflows, so I'll close this issue. Please let me know if this isn't working with the dockerised workflows @andrewGhazi
